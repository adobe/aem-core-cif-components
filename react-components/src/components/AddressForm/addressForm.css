.root {
}

.heading {
    background-color: white;
    font-size: 0.875rem;
    font-weight: 600;
    grid-column-end: span 2;
    line-height: 1rem;
    padding: 1.5rem 0 1.25rem;
    text-align: center;
    text-transform: uppercase;
    top: 0;
    z-index: 1;
}

.body {
    align-content: start;
    animation-duration: 224ms;
    animation-iteration-count: 1;
    animation-name: enter;
    background-color: white;
    bottom: 5rem;
    box-shadow: 0 -1px rgb(var(--venia-border));
    display: grid;
    grid-gap: 0.5rem;
    left: 0;
    max-height: calc(100vh - 8.5rem);
    overflow: auto;
    padding: 0 1.5rem;
    right: 0;
}

.footer {
    align-items: center;
    background-color: white;
    display: grid;
    grid-auto-columns: min-content;
    grid-auto-flow: column;
    grid-gap: 0.75rem;
    height: 5rem;
    justify-content: center;
    justify-items: center;
    position: relative;
}

@keyframes enter {
    from {
        opacity: 0;
        transform: translate3d(0, 100%, 0);
    }
    to {
        opacity: 1;
        transform: translate3d(0, 0, 0);
    }
}

/* fields */

.button {
    text-transform: capitalize;
}

.textInput {
    background: white;
    border: 1px solid rgb(var(--venia-text-alt));
    border-radius: 2px;
    color: rgb(var(--venia-text));
    display: inline-flex;
    flex: 0 0 100%;
    font-size: 0.9375rem;
    height: 2.25rem;
    padding: calc(0.375rem - 1px) calc(0.625rem - 1px);
    width: 100%;
}

.textInput:focus {
    border-color: rgb(var(--venia-text));
    outline: 0 none;
}

.city,
.firstname,
.lastname,
.postcode,
.region_code,
.telephone {
    grid-column-end: span 1;
}

.email,
.street0 {
    grid-column-end: span 2;
}

.default_shipping {
    grid-column-end: span 2;
    text-align: left;
}

.validation {
    grid-column-end: span 2;
    line-height: normal;
}

.error {
    color: rgb(var(--venia-error));
    grid-column-end: span 2;
    line-height: normal;
}
