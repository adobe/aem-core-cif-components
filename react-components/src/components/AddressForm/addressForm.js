/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { useCallback, useMemo, useState } from 'react';
import { Form } from 'informed';
import { array, bool, func, object, shape, string } from 'prop-types';
import { useTranslation } from 'react-i18next';

import Button from '../Button';
import classes from './addressForm.css';
import { validateEmail, isRequired, hasLengthExactly, validateRegionCode } from '../../utils/formValidators';
import combine from '../../utils/combineValidators';
import TextInput from '../TextInput';
import Field from '../Field';
import Checkbox from '../Checkbox';

const fields = [
    'city',
    'default_shipping',
    'email',
    'firstname',
    'lastname',
    'postcode',
    'region_code',
    'region',
    'street',
    'telephone'
];

const AddressForm = props => {
    const [submitting, setIsSubmitting] = useState(false);
    const {
        cancel,
        countries,
        formErrorMessage,
        heading,
        isAddressInvalid,
        invalidAddressMessage,
        initialValues,
        showDefaultAddressCheckbox,
        showEmailInput,
        submit,
        submitLabel
    } = props;
    const validationMessage = isAddressInvalid ? invalidAddressMessage : null;
    const errorMessage = formErrorMessage ? formErrorMessage : null;
    const [t] = useTranslation(['account', 'checkout', 'common']);

    const values = useMemo(
        () =>
            fields.reduce((acc, key) => {
                if (initialValues && key in initialValues) {
                    // Convert street from array to flat strings
                    if (key === 'street') {
                        initialValues[key].forEach((v, i) => (acc[`street${i}`] = v));
                        return acc;
                    }
                    // Convert region from object to region_code string
                    if (key === 'region') {
                        acc['region_code'] = initialValues[key].region_code;
                        return acc;
                    }
                    acc[key] = initialValues[key];
                }
                return acc;
            }, {}),
        [initialValues]
    );

    const handleSubmit = useCallback(
        values => {
            setIsSubmitting(true);
            // Convert street back to array
            submit({ ...values, street: [values.street0] });
        },
        [submit]
    );

    const submitButtonLabel = submitLabel || t('account:address-save', 'Save');
    const formHeading = heading || t('account:address-form-heading', 'Address');

    return (
        <Form className={classes.root} initialValues={values} onSubmit={handleSubmit}>
            <div className={classes.body}>
                <h2 className={classes.heading}>{formHeading}</h2>
                <div className={classes.firstname}>
                    <Field label={t('checkout:address-firstname', 'First Name')}>
                        <TextInput id={classes.firstname} field="firstname" validateOnBlur validate={isRequired} />
                    </Field>
                </div>
                <div className={classes.lastname}>
                    <Field label={t('checkout:address-lastname', 'Last Name')}>
                        <TextInput id={classes.lastname} field="lastname" validateOnBlur validate={isRequired} />
                    </Field>
                </div>
                {showEmailInput && (
                    <div className={classes.email}>
                        <Field label={t('checkout:address-email', 'E-Mail')}>
                            <TextInput
                                id={classes.email}
                                field="email"
                                validateOnBlur
                                validate={combine([isRequired, validateEmail])}
                            />
                        </Field>
                    </div>
                )}
                <div className={classes.street0}>
                    <Field label={t('checkout:address-street', 'Street')}>
                        <TextInput id={classes.street0} field="street0" validateOnBlur validate={isRequired} />
                    </Field>
                </div>
                <div className={classes.city}>
                    <Field label={t('checkout:address-city', 'City')}>
                        <TextInput id={classes.city} field="city" validateOnBlur validate={isRequired} />
                    </Field>
                </div>
                <div className={classes.region_code}>
                    <Field label={t('checkout:address-state', 'State')}>
                        <TextInput
                            id={classes.region_code}
                            field="region_code"
                            validateOnBlur
                            validate={combine([isRequired, [hasLengthExactly, 2], [validateRegionCode, countries]])}
                        />
                    </Field>
                </div>
                <div className={classes.postcode}>
                    <Field label={t('checkout:address-postcode', 'ZIP')}>
                        <TextInput id={classes.postcode} field="postcode" validateOnBlur validate={isRequired} />
                    </Field>
                </div>
                <div className={classes.telephone}>
                    <Field label={t('checkout:address-phone', 'Phone')}>
                        <TextInput id={classes.telephone} field="telephone" validateOnBlur validate={isRequired} />
                    </Field>
                </div>
                {showDefaultAddressCheckbox && (
                    <div className={classes.default_shipping}>
                        <Checkbox
                            id={classes.default_shipping}
                            label={t('checkout:address-default-address', 'Make my default address')}
                            field="default_shipping"
                        />
                    </div>
                )}
                <div className={classes.validation}>{validationMessage}</div>
                <div className={classes.error}>{errorMessage}</div>
            </div>
            <div className={classes.footer}>
                <Button onClick={cancel}>{t('common:cancel', 'Cancel')}</Button>
                <Button type="submit" priority="high" disabled={submitting}>
                    {submitButtonLabel}
                </Button>
            </div>
        </Form>
    );
};

AddressForm.propTypes = {
    cancel: func.isRequired,
    classes: shape({
        body: string,
        button: string,
        city: string,
        default_shipping: string,
        email: string,
        firstname: string,
        footer: string,
        heading: string,
        lastname: string,
        postcode: string,
        root: string,
        region_code: string,
        street0: string,
        telephone: string,
        validation: string
    }),
    countries: array,
    formErrorMessage: string,
    heading: string,
    invalidAddressMessage: string,
    initialValues: object,
    isAddressInvalid: bool,
    showDefaultAddressCheckbox: bool,
    showEmailInput: bool,
    submit: func.isRequired,
    submitting: bool,
    submitLabel: string
};

AddressForm.defaultProps = {
    initialValues: {}
};

export default AddressForm;

/*
const mockAddress = {
    country_id: 'US',
    firstname: 'Veronica',
    lastname: 'Costello',
    street: ['6146 Honey Bluff Parkway'],
    city: 'Calder',
    postcode: '49628-7978',
    region_code: 'MI',
    region: {
        region_code: 'MI'
    },
    telephone: '(555) 229-3326',
    email: 'veronica@example.com'
};
*/
