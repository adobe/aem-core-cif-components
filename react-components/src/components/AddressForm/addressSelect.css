.root {
    grid-column-end: span 2;
}
