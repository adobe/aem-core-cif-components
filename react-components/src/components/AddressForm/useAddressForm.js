/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useState } from 'react';
import { useUserContext } from '../../context/UserContext';
import { useMutation } from '@apollo/client';

import { useCountries } from '../../utils/hooks';
import { createAddress, updateAddress } from '../../actions/user';

import MUTATION_UPDATE_CUSTOMER_ADDRESS from '../../queries/mutation_update_customer_address.graphql';
import MUTATION_CREATE_CUSTOMER_ADDRESS from '../../queries/mutation_create_customer_address.graphql';

export const useAddressForm = () => {
    const [userState, { dispatch }] = useUserContext();
    const { countries } = useCountries();
    const [inProgress, setInProgress] = useState(false);

    const [createCustomerAddress] = useMutation(MUTATION_CREATE_CUSTOMER_ADDRESS);
    const [updateCustomerAddress] = useMutation(MUTATION_UPDATE_CUSTOMER_ADDRESS);

    let errorMessage = '';
    if (userState.addressFormError && userState.addressFormError.length > 0) {
        errorMessage = userState.addressFormError;
    }

    const findSavedAddress = address => {
        return address ? userState.currentUser.addresses.find(item => isSameAddress(item, address)) : null;
    };

    const getNewAddress = () => {
        return {
            city: '',
            firstname: '',
            lastname: '',
            postcode: '',
            region_code: '',
            save_in_address_book: false,
            street0: '',
            telephone: ''
        };
    };

    const getRegionCode = address => {
        return (address.region && (address.region.region_code || address.region.code)) || address.region_code;
    };

    const getRegionId = (countries, countryCode, regionCode) => {
        const region =
            countries &&
            countries
                .filter(country => country.id == countryCode && country.available_regions)
                .map(country => country.available_regions.find(region => region.code == regionCode));
        return region && region[0] ? region[0].id : null;
    };

    const handleSubmit = async formValues => {
        setInProgress(true);

        const variables = {
            country_code: 'US',
            region: {
                region_code: formValues.region_code,
                region_id: getRegionId(countries, 'US', formValues.region_code)
            },
            default_billing: formValues.default_shipping,
            ...formValues
        };

        let resetFields;
        if (formValues.default_shipping) {
            resetFields = {
                default_shipping: false,
                default_billing: false
            };
        }

        if (userState.updateAddress) {
            variables.id = userState.updateAddress.id;
            updateAddress({ updateCustomerAddress, variables, resetFields, dispatch });
        } else {
            createAddress({ createCustomerAddress, variables, resetFields, dispatch });
        }

        setInProgress(false);
    };

    const handleCancel = () => {
        dispatch({ type: 'clearAddressFormError' });
        if (userState.updateAddress) {
            dispatch({ type: 'endEditingAddress' });
        }
        dispatch({ type: 'closeAddressForm' });
    };

    const isSameAddress = (address1, address2) => {
        return (
            address1.firstname === address2.firstname &&
            address1.lastname === address2.lastname &&
            address1.street.join(' ') === address2.street.join(' ') &&
            address1.city === address2.city &&
            address1.country_code === address2.country_code &&
            getRegionCode(address1) === getRegionCode(address2) &&
            address1.postcode === address2.postcode &&
            address1.telephone === address2.telephone
        );
    };

    const parseAddress = (address, email) => {
        let result = {
            ...address,
            region_code: getRegionCode(address),
            country_code: address.country_code || address.country.code
        };
        if (email) {
            result.email = email;
        }
        return result;
    };

    const parseAddressFormValues = address => {
        const fields = [
            'city',
            'default_shipping',
            'email',
            'firstname',
            'lastname',
            'postcode',
            'region_code',
            'region',
            'save_in_address_book',
            'street',
            'telephone'
        ];

        return fields.reduce((acc, key) => {
            if (address && key in address) {
                // Convert street from array to flat strings
                if (key === 'street') {
                    address[key].forEach((v, i) => (acc[`street${i}`] = v));
                    return acc;
                }
                // Convert region from object to region_code string, region object returned in different graphql
                // endpoints has different shape, so we have to check both 'region_code' and 'code' to get the
                // value of the region code
                if (key === 'region') {
                    acc['region_code'] = address[key].region_code || address[key].code;
                    return acc;
                }
                acc[key] = address[key];
            }
            return acc;
        }, {});
    };

    return {
        countries,
        errorMessage,
        findSavedAddress,
        getNewAddress,
        getRegionCode,
        getRegionId,
        handleSubmit,
        handleCancel,
        inProgress,
        isSameAddress,
        parseAddress,
        parseAddressFormValues,
        updateAddress: userState.updateAddress
    };
};
