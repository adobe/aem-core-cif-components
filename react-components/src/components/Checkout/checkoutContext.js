/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React, { createContext, useContext, useReducer } from 'react';
import { string, func, shape, object } from 'prop-types';

export const initialCheckoutState = {
    flowState: 'cart',
    order: null,
    editing: null,
    shippingAddress: null,
    billingAddress: null,
    shippingMethod: null,
    paymentMethod: null,
    braintreeToken: false
};

export const checkoutReducer = (state, action) => {
    switch (action.type) {
        case 'beginCheckout':
            return {
                ...state,
                flowState: 'form'
            };
        case 'cancelCheckout':
            return {
                ...state,
                flowState: 'cart'
            };
        case 'placeOrder':
            return {
                ...state,
                order: action.order,
                flowState: 'receipt',
                editing: 'receipt'
            };
        case 'reset':
            return {
                ...initialCheckoutState
            };
        case 'setEditing':
            return {
                ...state,
                editing: action.editing
            };
        case 'endEditing':
            return {
                ...state,
                editing: null
            };
        case 'setShippingAddress':
            return {
                ...state,
                shippingAddress: action.shippingAddress,
                editing: null
            };
        case 'setBillingAddress':
            return {
                ...state,
                billingAddress: action.billingAddress,
                editing: null
            };
        case 'setShippingMethod':
            return {
                ...state,
                shippingMethod: action.shippingMethod,
                editing: null
            };
        case 'setPaymentMethod':
            return {
                ...state,
                paymentMethod: action.paymentMethod,
                editing: null
            };
        case 'setBraintreeToken':
            return {
                ...state,
                braintreeToken: action.token
            };

        default:
            return state;
    }
};

export const CheckoutContext = createContext();

export const CheckoutProvider = props => {
    const reducer = props.reducer || checkoutReducer;
    const initialState = props.initialState || initialCheckoutState;
    return (
        <CheckoutContext.Provider value={useReducer(reducer, initialState)}>{props.children}</CheckoutContext.Provider>
    );
};

CheckoutProvider.propTypes = {
    reducer: func,
    initialState: shape({
        flowState: string.isRequired,
        order: object,
        editing: object,
        shippingAddress: object,
        billingAddress: object,
        shippingMethod: object,
        paymentMethod: object
    })
};

export const useCheckoutState = () => useContext(CheckoutContext);
