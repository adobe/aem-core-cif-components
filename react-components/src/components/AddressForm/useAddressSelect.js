/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useAddressForm } from './useAddressForm';
import { useCheckoutState } from '../Checkout/checkoutContext';
import { useUserContext } from '../../context/UserContext';

export const useAddressSelect = (props = { initialAddress: undefined }) => {
    const { initialAddress } = props;

    const { findSavedAddress, getNewAddress, parseAddress, parseAddressFormValues } = useAddressForm();
    const [, dispatch] = useCheckoutState();
    const [{ currentUser }] = useUserContext();

    const parseInitialAddressSelectValue = address => {
        let initialValue = 0; // '0' is the value of option item 'New Address' in address select
        const foundAddress = findSavedAddress(address);
        return foundAddress ? foundAddress.id : initialValue;
    };
    const initialAddressId = initialAddress ? parseInitialAddressSelectValue(initialAddress) : 0;
    const [selectedAddressId, setSelectedAddressId] = useState(initialAddressId);

    const [t] = useTranslation('checkout');

    const addressSelectNewAddressItem = {
        label: t('checkout:address-form-address-select-new-address', 'New Address'),
        value: 0
    };
    const addressSelectItems = currentUser.addresses.map(address => {
        return {
            label: address.street.join(' '),
            value: address.id
        };
    });
    addressSelectItems.unshift(addressSelectNewAddressItem);

    const handleChangeAddressSelectInCheckout = (value, addressFormApi) => {
        const newAddressItemValue = 0;
        setSelectedAddressId(value);
        if (newAddressItemValue == value) {
            // clear the values of the form fields if the current select option item is 'New Address'
            addressFormApi.setValues(getNewAddress());
            dispatch({ type: 'setIsEditingNewAddress', editing: true });
            return;
        }

        const address = currentUser.addresses.find(address => address.id == value);
        // parse the "GraphQL format" address and normalize the properties (i.e. add the region_code on the same level as the rest of the props)
        const parsedAddress = parseAddress(address);
        // update form values by using formApi passed from <AddressForm> component, this is due to the fact that
        // this is the proposed way to update values of fields in 'informed' form library
        const parsedFormValues = parseAddressFormValues(parsedAddress);
        addressFormApi.setValues(parsedFormValues);
    };

    return {
        addressSelectItems,
        selectedAddressId,
        handleChangeAddressSelectInCheckout,
        parseInitialAddressSelectValue
    };
};
