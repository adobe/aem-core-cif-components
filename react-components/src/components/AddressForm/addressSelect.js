/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { func, number } from 'prop-types';

import { useAddressSelect } from './useAddressSelect';
import Select from '../Select';

import classes from './addressSelect.css';

const AddressSelect = props => {
    const { initialValue, onValueChange } = props;
    const { addressSelectItems } = useAddressSelect();

    return (
        <div className={classes.root}>
            <Select
                items={addressSelectItems}
                field="address_select"
                initialValue={initialValue}
                onValueChange={onValueChange}
            />
        </div>
    );
};

AddressSelect.propTypes = {
    initialValue: number,
    onValueChange: func
};

export default AddressSelect;
