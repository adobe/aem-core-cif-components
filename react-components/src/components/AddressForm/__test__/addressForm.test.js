/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { screen, fireEvent, wait } from '@testing-library/react';
import { render } from 'test-utils';
import { CheckoutProvider } from '../../Checkout';
import AddressForm from '../addressForm';

describe('<AddressForm />', () => {
    const countries = [
        {
            id: 'US',
            available_regions: [
                {
                    name: 'Michigan',
                    code: 'MI'
                }
            ]
        }
    ];

    it('renders the component', async () => {
        const { asFragment } = render(<AddressForm cancel={() => {}} submit={() => {}} />);
        await wait(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });

    it('renders the component with address validation message and form error message', async () => {
        const { asFragment } = render(
            <AddressForm
                isAddressInvalid={true}
                validationMessage={'address validation message'}
                formErrorMessage={'form error message'}
                cancel={() => {}}
                submit={() => {}}
            />
        );
        await wait(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });

    it('renders the component with address select', async () => {
        const { asFragment } = render(
            <CheckoutProvider>
                <AddressForm
                    showAddressSelect={true}
                    initialAddressSelectValue={0}
                    onAddressSelectValueChange={() => {}}
                    cancel={() => {}}
                    submit={() => {}}
                />
            </CheckoutProvider>
        );
        await wait(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });

    it('renders the component with default address checkbox', async () => {
        const { asFragment } = render(
            <CheckoutProvider>
                <AddressForm showDefaultAddressCheckbox={true} cancel={() => {}} submit={() => {}} />
            </CheckoutProvider>
        );
        await wait(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });

    it('renders the component with save in address book checkbox', async () => {
        const { asFragment } = render(
            <CheckoutProvider>
                <AddressForm showSaveInAddressBookCheckbox={true} cancel={() => {}} submit={() => {}} />
            </CheckoutProvider>
        );
        await wait(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });

    it('calls the callback function when changing the address select', async () => {
        const onAddressSelectValueChange = jest.fn(() => {});
        render(
            <CheckoutProvider>
                <AddressForm
                    showAddressSelect={true}
                    initialAddressSelectValue={0}
                    onAddressSelectValueChange={onAddressSelectValueChange}
                    cancel={() => {}}
                    submit={() => {}}
                />
            </CheckoutProvider>
        );
        const addressSelect = screen.getByDisplayValue('New Address');
        expect(addressSelect).not.toBeUndefined();
        fireEvent.change(addressSelect, { target: { value: '1' } });
        await wait(() => {
            expect(onAddressSelectValueChange).toHaveBeenCalledTimes(1);
        });
    });

    it('fills the street0 field with a given street array value', async () => {
        const initialValues = {
            street: ['street A', 'street B']
        };

        const { container } = render(<AddressForm cancel={() => {}} submit={() => {}} initialValues={initialValues} />);

        await wait(() => {
            const streetField = container.querySelector('#street0');
            expect(streetField.value).toEqual('street A');
        });
    });

    it('returns the street field as array', async () => {
        const initialValues = {
            country_id: 'US',
            firstname: 'Veronica',
            lastname: 'Costello',
            city: 'Calder',
            postcode: '49628-7978',
            region_code: 'MI',
            region: {
                region_code: 'MI'
            },
            telephone: '(555) 229-3326',
            email: 'veronica@example.com'
        };

        const mockSubmit = jest.fn(() => {});
        const { container } = render(
            <AddressForm cancel={() => {}} submit={mockSubmit} initialValues={initialValues} countries={countries} />
        );

        // Fill street input
        const streetField = container.querySelector('#street0');
        fireEvent.change(streetField, { target: { value: 'street A' } });

        // Click on submit
        let submit = container.querySelector('button[type=submit]');
        fireEvent.click(submit);

        // Verify parameter
        await wait(() => {
            expect(mockSubmit).toHaveBeenCalledTimes(1);
            const submitValues = mockSubmit.mock.calls[0][0];
            expect(submitValues.street).toEqual(['street A']);
        });
    });
});
