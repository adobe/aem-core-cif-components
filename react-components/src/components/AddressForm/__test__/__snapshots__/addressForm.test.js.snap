// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressForm /> renders the component 1`] = `
<DocumentFragment>
  <form
    class="root"
  >
    <div
      class="body"
    >
      <h2
        class="heading"
      >
        Address
      </h2>
      <div
        class="firstname"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            First Name
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="firstname"
                name="firstname"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="lastname"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            Last Name
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="lastname"
                name="lastname"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="street0"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            Street
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="street0"
                name="street0"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="city"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            City
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="city"
                name="city"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="region_code"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            State
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="region_code"
                name="region_code"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="postcode"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            ZIP
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="postcode"
                name="postcode"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="telephone"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            Phone
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="telephone"
                name="telephone"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="validation"
      />
      <div
        class="error"
      />
    </div>
    <div
      class="footer"
    >
      <button
        class="root_normalPriority"
        type="button"
      >
        <span
          class="content"
        >
          Cancel
        </span>
      </button>
      <button
        class="root_highPriority"
        type="submit"
      >
        <span
          class="content"
        >
          Save
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<AddressForm /> renders the component with address validation message and form error message 1`] = `
<DocumentFragment>
  <form
    class="root"
  >
    <div
      class="body"
    >
      <h2
        class="heading"
      >
        Address
      </h2>
      <div
        class="firstname"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            First Name
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="firstname"
                name="firstname"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="lastname"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            Last Name
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="lastname"
                name="lastname"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="street0"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            Street
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="street0"
                name="street0"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="city"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            City
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="city"
                name="city"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="region_code"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            State
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="region_code"
                name="region_code"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="postcode"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            ZIP
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="postcode"
                name="postcode"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="telephone"
      >
        <div
          class="root"
        >
          <span
            class="label"
          >
            Phone
          </span>
          <span
            class="root"
          >
            <span
              class="input"
            >
              <input
                class="input"
                id="telephone"
                name="telephone"
                value=""
              />
            </span>
            <span
              class="before"
            />
            <span
              class="after"
            />
          </span>
          <p
            class="root"
          />
        </div>
      </div>
      <div
        class="validation"
      />
      <div
        class="error"
      >
        form error message
      </div>
    </div>
    <div
      class="footer"
    >
      <button
        class="root_normalPriority"
        type="button"
      >
        <span
          class="content"
        >
          Cancel
        </span>
      </button>
      <button
        class="root_highPriority"
        type="submit"
      >
        <span
          class="content"
        >
          Save
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
